src/comparator.rs

```diff
@@ -1,30 +1,482 @@
+
+positive_input_pin!(COMP2, gpiob::PB4<Analog>, 0b00);
+positive_input_pin!(COMP2, gpiob::PB6<Analog>, 0b01);
+positive_input_pin!(COMP2, gpioa::PA3<Analog>, 0b10);
+positive_input_pin!(COMP2, Open, 0b11);
+
+macro_rules! negative_input_pin {
+    ($COMP:ident, $pin:ty, $bits:expr) => {
+        impl NegativeInput<$COMP> for $pin {
+            fn setup(&self, comp: &$COMP) {
+                comp.csr().modify(|_, w| unsafe { w.inmsel().bits($bits) })
+            }
+        }
+    };
+}
+
+negative_input_pin!(COMP1, gpiob::PB1<Analog>, 0b0110);
+negative_input_pin!(COMP1, gpioc::PC4<Analog>, 0b0111);
+negative_input_pin!(COMP1, gpioa::PA0<Analog>, 0b1000);
+
+negative_input_pin!(COMP2, gpiob::PB3<Analog>, 0b0110);
+negative_input_pin!(COMP2, gpiob::PB7<Analog>, 0b0111);
+negative_input_pin!(COMP2, gpioa::PA2<Analog>, 0b1000);
+
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub enum RefintInput {
+    /// VRefint * 1/4
+    VRefintM14 = 0b0000,
+    /// VRefint * 1/2
+    VRefintM12 = 0b0001,
+    /// VRefint * 3/4
+    VRefintM34 = 0b0010,
+    /// VRefint
+    VRefint = 0b0011,
+}
+
+macro_rules! refint_input {
+    ($COMP:ident) => {
+        impl NegativeInput<$COMP> for RefintInput {
+            fn setup(&self, comp: &$COMP) {
+                comp.csr()
+                    .modify(|_, w| unsafe { w.inmsel().bits(*self as u8) })
+            }
+        }
+    };
+}
+
+refint_input!(COMP1);
+refint_input!(COMP2);
+
+macro_rules! dac_input {
+    ($COMP:ident, $channel:ty, $bits:expr) => {
+        impl NegativeInput<$COMP> for &$channel {
+            fn setup(&self, comp: &$COMP) {
+                comp.csr().modify(|_, w| unsafe { w.inmsel().bits($bits) })
+            }
+        }
+    };
+}
+
+#[cfg(any(feature = "stm32g071", feature = "stm32g081"))]
+dac_input!(COMP1, dac::Channel1<dac::Enabled>, 0b0100);
+#[cfg(any(feature = "stm32g071", feature = "stm32g081"))]
+dac_input!(COMP1, dac::Channel2<dac::Enabled>, 0b0101);
+
+#[cfg(any(feature = "stm32g071", feature = "stm32g081"))]
+dac_input!(COMP2, dac::Channel1<dac::Enabled>, 0b0100);
+#[cfg(any(feature = "stm32g071", feature = "stm32g081"))]
+dac_input!(COMP2, dac::Channel2<dac::Enabled>, 0b0101);
+
+pub struct Comparator<C> {
+    regs: C,
+}
+
+pub trait ComparatorExt<COMP> {
+    fn init<P: PositiveInput<COMP>, N: NegativeInput<COMP>>(
+        &mut self,
+        positive_input: P,
+        negative_input: N,
+        config: Config,
+    );
+    fn output(&self) -> bool;
+    fn enable(&self);
+    fn disable(&self);
+    fn output_pin<P: OutputPin<COMP>>(&self, pin: P);
+    fn listen(&self, edge: SignalEdge, exti: &EXTI);
+    fn unlisten(&self, exti: &EXTI);
+    fn is_pending(&self, edge: SignalEdge, exti: &EXTI) -> bool;
+    fn unpend(&self, exti: &EXTI);
+}
+
+macro_rules! comparator_ext {
+    ($COMP:ty, $Comparator:ty, $Event:expr) => {
+        impl ComparatorExt<$COMP> for $Comparator {
+            // TODO: Require init before any other functions?
```

Comment:
It would be excellent if this is made mandatory by type system, so it would be impossible to call wrong methods if peripheral is not properly setup.

Reply:
I changed the `Ext` traits to work more like the other peripherals, and initialization is forced before any other function is called.